python/jupyterlab-chat/jupyterlab_chat/ychat.py

```diff
@@ -168,24 +168,43 @@ def update_message(self, message: Message, append: bool = False):
                 message.body = initial_message["body"] + message.body  # type:ignore[index]
             self._ymessages[index] = asdict(message, dict_factory=message_asdict_factory)
 
-    def get_attachments(self) -> dict[str, Attachment]:
+    def get_attachments(self) -> dict[str, Union[FileAttachment, CellAttachment]]:
         """
-        Returns the attachments of the document.
+        Returns all attachments in the chat as a dictionary, indexed by
+        attachment ID.
         """
         return self._yattachments.to_py() or {}
 
-    def set_attachment(self, attachment: Attachment):
-        """
-        Add or modify an attachments of the document.
-        """
-        attachment_id = str(uuid4())
-        with self._ydoc.transaction():
-            # Search if the attachment already exist to update it, otherwise add it.
+    def set_attachment(self, attachment: Union[FileAttachment, CellAttachment]) -> str:
+        """
+        Add or modify an attachment in the chat, and returns the ID of the
+        attachment.
+
+        NOTE: This method does not add an attachment to any message. It merely
+        adds the attachment data to the chat file and returns an attachment ID.
+        To add an attachment to a new message, consumers should call this method
+        & add the returned ID to `NewMessage.attachments`.
+        """
+        if attachment.selection:
+            # Generate a unique ID if the attachment contains a selection.
+            # Attachments with selections are always considered unique because
+            # the selection range is generally specific.
+            attachment_id = str(uuid4())
+        else:
+            # Otherwise, use the ID of the existing attachment if one exists.
+            # Attachments without selections are considered identical if they
+            # share the same `type` and `value`. 
             for id, att in self.get_attachments().items():
                 if att.type == attachment.type and att.value == attachment.value:
```

Comment:
Same as above, if the python representation of the map object is consistent, we could compare the JSON value.